### PR TITLE
Wire advanced audit to asset and oauth api servers

### PR DESCRIPTION
**Pull request by soltysh**
@deads2k ptal

### CODE REVIEW

**Comment by soltysh** [comment]
[(3) [SCCFSI] audit log needs to capture all login events without turning on debug levels](https://trello.com/c/d4v3ZEfo/944-3-sccfsi-audit-log-needs-to-capture-all-login-events-without-turning-on-debug-levels)

File: pkg/cmd/server/origin/asset_apiserver_adapter.go [hunk 1]
```diff
@@ -25,6 +26,13 @@ func NewAssetServerConfigFromMasterConfig(masterConfigOptions configapi.MasterCo
 		return nil, err
 	}
 
+	backend, policy, err := kubernetes.GetAuditConfig(masterConfigOptions.AuditConfig)
```
Comment (deads2k) [inline pkg/cmd/server/origin/asset_apiserver_adapter.go]: Find the call site of `NewAssetServerConfigFromMasterConfig` and set the backend and policy there.

File: pkg/assets/apiserver/asset_apiserver.go [hunk 2]
```diff
@@ -46,6 +50,9 @@ type AssetServerConfig struct {
 	LimitRequestOverrides *clusterresourceoverrideapi.ClusterResourceOverrideConfig
 
 	PublicURL url.URL
+
+	AuditBackend       audit.Backend
```
Comment (deads2k) [inline pkg/assets/apiserver/asset_apiserver.go]: use the ones already contained in `GenericConfig`

File: pkg/auth/oauth/handlers/default_auth_handler.go [hunk 3]
```diff
@@ -7,21 +7,23 @@ import (
...
+	challengers          map[string]AuthenticationChallenger
+	redirectors          *AuthenticationRedirectors
+	errorHandler         AuthenticationErrorHandler
+	selectionHandler     AuthenticationSelectionHandler
+	requestContextMapper request.RequestContextMapper
 }
 
 // NewUnionAuthenticationHandler returns an oauth.AuthenticationHandler that muxes multiple challenge handlers and redirect handlers
-func NewUnionAuthenticationHandler(passedChallengers map[string]AuthenticationChallenger, passedRedirectors *AuthenticationRedirectors, errorHandler AuthenticationErrorHandler, selectionHandler AuthenticationSelectionHandler) AuthenticationHandler {
+func NewUnionAuthenticationHandler(passedChallengers map[string]AuthenticationChallenger, passedRedirectors *AuthenticationRedirectors, errorHandler AuthenticationErrorHandler, selectionHandler AuthenticationSelectionHandler, requestContextMapper request.RequestContextMapper) AuthenticationHandler {
```
Comment (deads2k) [inline pkg/auth/oauth/handlers/default_auth_handler.go]: @openshift/sig-security ptal

File: pkg/cmd/server/origin/oauth_apiserver_adapter.go [hunk 4]
```diff
@@ -69,5 +70,12 @@ func NewOAuthServerConfigFromMasterConfig(masterConfig *MasterConfig) (*oauthapi
 		oauthServerConfig.AssetPublicAddresses = []string{oauthConfig.AssetPublicURL}
 	}
 
+	backend, policy, err := kubernetes.GetAuditConfig(masterConfig.Options.AuditConfig)
```
Comment (deads2k) [inline pkg/cmd/server/origin/oauth_apiserver_adapter.go]: same bit about finding callsites for this guy and overriding post construction.

File: pkg/oauth/apiserver/oauth_apiserver.go [hunk 5]
```diff
@@ -55,6 +59,9 @@ type OAuthServerConfig struct {
 	SessionAuth *session.Authenticator
 
 	HandlerWrapper handlerWrapper
+
+	AuditBackend       audit.Backend
```
Comment (deads2k) [inline pkg/oauth/apiserver/oauth_apiserver.go]: use the ones from the genericconfig

**Comment by soltysh** [comment]
Rebased and comments addressed. ptal

**Comment by soltysh** [comment]
@mfojtik fyi

File: pkg/cmd/server/kubernetes/master/master_config.go [hunk 6]
```diff
@@ -456,52 +456,12 @@ func buildKubeApiserverConfig(
...
-				webhook, err := auditwebhook.NewBackend(masterConfig.AuditConfig.WebHookKubeConfig, string(masterConfig.AuditConfig.WebHookMode))
-				if err != nil {
-					glog.Fatalf("Audit webhook initialization failed: %v", err)
-				}
-				genericConfig.AuditBackend = audit.Union(genericConfig.AuditBackend, webhook)
-			}
-		}
+	backend, policyChecker, err := GetAuditConfig(masterConfig.AuditConfig)
+	if err != nil {
+		glog.Fatalf("Audit initialization failed: %v", err)
```
Comment (deads2k) [inline pkg/cmd/server/kubernetes/master/master_config.go]: method returns an error

File: pkg/cmd/server/origin/master.go [hunk 7]
```diff
@@ -162,6 +162,12 @@ func (c *MasterConfig) newAssetServerHandler() (http.Handler, error) {
 	if err != nil {
 		return nil, err
 	}
+	backend, policy, err := kubernetes.GetAuditConfig(c.Options.AuditConfig)
```
Comment (deads2k) [inline pkg/cmd/server/origin/master.go]: This method is called from a method with access to the genericapiserver config.  Pass that config through to here and assign it.  We probably want all apiservers to use the same audit config and that is the source of truth at this level.

File: pkg/auth/oauth/handlers/default_auth_handler.go (hunk 3 above)
Comment (deads2k) [reply pkg/auth/oauth/handlers/default_auth_handler.go]: @simo5 @enj The audit wiring here will end up with you.  You ought to take a look.

**Comment by soltysh** [comment]
@deads2k comments addressed

File: pkg/cmd/server/start/start_master.go [hunk 8]
```diff
@@ -584,6 +584,8 @@ func StartAPI(oc *origin.MasterConfig, kubeAPIServerConfig *master.Config, infor
 		if err != nil {
 			return err
 		}
+		config.GenericConfig.AuditBackend = kubeAPIServerConfig.GenericConfig.AuditBackend
```
Comment (deads2k) [inline pkg/cmd/server/start/start_master.go]: This was a good usage of your `GetAuditConfig` bit.  The `kubeapiserverconfig` goes away here: https://github.com/openshift/origin/pull/16582

Comment (soltysh) [reply pkg/cmd/server/start/start_master.go]: sigh... I'll revert it back

File: pkg/auth/oauth/handlers/default_auth_handler.go [hunk 9]
```diff
@@ -113,6 +115,21 @@ func (authHandler *unionAuthenticationHandler) AuthenticationNeeded(apiClient au
 				w.WriteHeader(http.StatusFound)
 			default:
 				w.WriteHeader(http.StatusUnauthorized)
+				ctx, ok := authHandler.requestContextMapper.Get(req)
+				if !ok {
+					return false, fmt.Errorf("no context found for request to audit")
```
Comment (enj) [inline pkg/auth/oauth/handlers/default_auth_handler.go]: Why does this error?

File: pkg/auth/oauth/handlers/default_auth_handler.go [hunk 10]
```diff
@@ -210,3 +227,33 @@ func mergeHeaders(dest http.Header, toAdd http.Header) {
...
+		authMethods = append(authMethods, "basic")
+	}
+
+	auth := strings.TrimSpace(req.Header.Get("Authorization"))
+	parts := strings.Split(auth, " ")
+	if len(parts) > 1 && strings.ToLower(parts[0]) == "bearer" {
+		authMethods = append(authMethods, "bearer")
+	}
+
+	token := strings.TrimSpace(req.URL.Query().Get("access_token"))
```
Comment (enj) [inline pkg/auth/oauth/handlers/default_auth_handler.go]: No error and code params? They are also part of the OAuth flow.

File: pkg/auth/oauth/handlers/default_auth_handler.go [hunk 11]
```diff
@@ -113,6 +115,21 @@ func (authHandler *unionAuthenticationHandler) AuthenticationNeeded(apiClient au
...
+				}
+				ev := request.AuditEventFrom(ctx)
+				if ev != nil {
+					// this code mimics the bits from k8s.io/apiserver/pkg/endpoints/filters/authn_audit.go
+					// but since we don't accept failedHander here we need to manually alter the audit
+					// event with information about failed authentication
+					ev.ResponseStatus.Message = getAuthMethods(req)
+					ctx = request.WithAuditEvent(ctx, ev)
+					if err := authHandler.requestContextMapper.Update(req, ctx); err != nil {
+						return false, fmt.Errorf("failed to attach audit event to context: %v", err)
```
Comment (enj) [inline pkg/auth/oauth/handlers/default_auth_handler.go]: Why does this error?

File: pkg/auth/oauth/handlers/default_auth_handler.go [hunk 12]
```diff
@@ -113,6 +115,21 @@ func (authHandler *unionAuthenticationHandler) AuthenticationNeeded(apiClient au
 				w.WriteHeader(http.StatusFound)
 			default:
 				w.WriteHeader(http.StatusUnauthorized)
+				ctx, ok := authHandler.requestContextMapper.Get(req)
+				if !ok {
+					return false, fmt.Errorf("no context found for request to audit")
+				}
+				ev := request.AuditEventFrom(ctx)
+				if ev != nil {
```
Comment (enj) [inline pkg/auth/oauth/handlers/default_auth_handler.go]: I assume we do not care about the `ev == nil` case?

File: pkg/cmd/server/kubernetes/master/master_config.go (hunk 6 above)
Comment (enj) [reply pkg/cmd/server/kubernetes/master/master_config.go]: @soltysh ?

**Comment by enj** [comment]
Not sure why this is a critical fix.  I trust you guys so I will not try to block it.  @soltysh PTAL at my comments.

File: pkg/auth/oauth/handlers/default_auth_handler.go (hunk 9 above)
Comment (soltysh) [reply pkg/auth/oauth/handlers/default_auth_handler.go]: It's very serious error if a request does not have associated context, thus we'll error here.

File: pkg/auth/oauth/handlers/default_auth_handler.go (hunk 12 above)
Comment (soltysh) [reply pkg/auth/oauth/handlers/default_auth_handler.go]: Nope, we don't. This means audit is not happening and there's nothing we need to do here.

File: pkg/auth/oauth/handlers/default_auth_handler.go (hunk 11 above)
Comment (soltysh) [reply pkg/auth/oauth/handlers/default_auth_handler.go]: Similar to above case where we get context, after we modify the context with new event we need to post that back. If that fails, we're loosing information added here.

File: pkg/auth/oauth/handlers/default_auth_handler.go (hunk 10 above)
Comment (soltysh) [reply pkg/auth/oauth/handlers/default_auth_handler.go]: That was discussed with @liggitt in the original PR (https://github.com/kubernetes/kubernetes/pull/51119). We want to provide with enough information that auth failed and the method we've tried but without going into too much details.

**Comment by soltysh** [comment]
> Not sure why this is a critical fix. I trust you guys so I will not try to block it. @soltysh PTAL at my comments.

All in all that piece of code is a nasty duplication from [audit_authn.go](https://github.com/openshift/origin/blob/b7cc46fc1342250dfd0e115a37af853496e59b88/vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/endpoints/filters/authn_audit.go#L61) and [audit.go](https://github.com/openshift/origin/blob/b7cc46fc1342250dfd0e115a37af853496e59b88/vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go#L113). The reason for doing it this way is that our oauth flow is different than the one in k8s and I had to manually weave this in. 
Having said that, I'm open to suggestion how we can improve that :-)

File: pkg/cmd/server/kubernetes/master/master_config.go (hunk 6 above)
Comment (soltysh) [reply pkg/cmd/server/kubernetes/master/master_config.go]: I swear I fixed that. Must have been lost when going back and forth on this piece. I've opened https://github.com/openshift/origin/pull/16594 to address that.

**Pull request merged**